### PR TITLE
Support eval mode for dynamicemb

### PR DESCRIPTION
**Pull request by shijieliu**
## Checklist
- [ ] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/rapids-repo-template/blob/HEAD/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

File: corelib/dynamicemb/dynamicemb/batched_dynamicemb_function.py [hunk 1]
```diff
@@ -114,13 +116,23 @@ def forward(
                 num_unique_indices, dims[i], dtype=tmp_value_type_torch, device=device
```
Comment (jiashuy) [inline corelib/dynamicemb/dynamicemb/batched_dynamicemb_function.py:122]: using torch.zeros here

File: corelib/dynamicemb/dynamicemb/batched_dynamicemb_function.py [hunk 2]
```diff
@@ -114,13 +116,23 @@ def forward(
...
+            if training:
+                find_or_insert(
+                    tables[i],
+                    num_unique_indices,
+                    unique_indices,
+                    tmp_unique_embs,
+                    scores[i],
+                )
+            else:
+                tmp_founds = torch.empty(num_unique_indices, dtype=torch.bool, device=device)
```
Comment (jiashuy) [inline corelib/dynamicemb/dynamicemb/batched_dynamicemb_function.py]: We do not support find() for concating embedding and optimizer states.

File: corelib/dynamicemb/src/dynamic_emb_op.cu [hunk 3]
```diff
@@ -464,35 +490,46 @@ void lookup_forward_dense(
...
+          auto score = std::make_optional<uint64_t>(py::cast<uint64_t>(scores[i]));
+          find_or_insert(tables[i], tmp_unique_num, tmp_unique_indices[i],
+                        tmp_unique_embs, score);
+        }
       } else {
-        auto score = std::make_optional<uint64_t>(py::cast<uint64_t>(scores[i]));
-        find_or_insert(tables[i], tmp_unique_num, tmp_unique_indices[i],
-                      tmp_unique_embs, score);
+        at::Tensor tmp_founds = create_sub_tensor(founds, unique_embs_offset);
+        find(tables[i], tmp_unique_num, tmp_unique_indices[i], tmp_unique_embs, tmp_founds);
```
Comment (jiashuy) [inline corelib/dynamicemb/src/dynamic_emb_op.cu]: Don't support find embedding from concated(embedding, optimizer states) table.

**Review by jiashuy — COMMENTED** [review]
Need to develop find, to support only lookup embedding based on main branch.

**Comment by WhoisZihan** [comment]
I've modified with the latest interface, I can refresh the commit later.

File: corelib/dynamicemb/test/test_batched_dynamic_embedding_tables.py [hunk 4]
```diff
@@ -80,20 +101,98 @@ def test_embedding_optimizer(opt_type, opt_params):
...
+    f3  [],         [0]
+    """
+    indices = torch.tensor(
+        [0, 1, 12, 64, 8, 12, 15, 2, 7, 105, 0], dtype=key_type, device=device
+    )
+    offsets = torch.tensor(
+        [0, 2, 3, 5, 6, 8, 10, 10, 11], dtype=key_type, device=device
+    )
+
+    embs_train = bdebt(indices, offsets)
```
Comment (jiashuy) [inline corelib/dynamicemb/test/test_batched_dynamic_embedding_tables.py:171]: So by default, nn.Module.training is ture?

File: corelib/dynamicemb/dynamicemb/batched_dynamicemb_tables.py [hunk 5]
```diff
@@ -650,6 +650,10 @@ def forward(
         # Jost forward it to DynamicEmbeddingFunction
         # return DynamicEmbeddingFunction.apply(indices, offsets, self.table_offsets_in_feature, self.tables, self.total_D,
         #                                       self.dims,self.feature_table_map, self.embedding_dtype, self.pooling_mode, torch.device(self.device_id), 1, self._empty_tensor)
+        if any([not o.training for o in self._dynamicemb_options]) and self.training:
```
Comment (jiashuy) [inline corelib/dynamicemb/dynamicemb/batched_dynamicemb_tables.py:653]: Maybe option.training should have a more suitable name. It just used in model initialization and device whether allocate memory for optimizer states.

File: corelib/dynamicemb/test/test_batched_dynamic_embedding_tables.py (hunk 4 above)
Comment (shijieliu) [reply corelib/dynamicemb/test/test_batched_dynamic_embedding_tables.py:171]: I think so.

**Comment by shijieliu** [comment]
hi @WhoisZihan could you help review this PR?

File: corelib/dynamicemb/dynamicemb/optimizer.py [hunk 6]
```diff
@@ -201,29 +197,15 @@ def update(
...
-                    ht,
-                    num_indice,
-                    indice,
-                    grad,
-                    lr,
-                    weight_dtype,
-                    score,
-                )
+
+            dynamic_emb_sgd_with_table(
```
Comment (WhoisZihan) [inline corelib/dynamicemb/dynamicemb/optimizer.py:201]: The optimizer change seems not relevant to eval mode, is it another feature?

Comment (shijieliu) [reply corelib/dynamicemb/dynamicemb/optimizer.py:201]: Yes. We move the insert in backward to forward when use_index_dedup=True. This will make the behavior of dynamicemb more consistent.

Comment (shijieliu) [reply corelib/dynamicemb/dynamicemb/optimizer.py:201]: Update the PR name

**Comment by WhoisZihan** [comment]
LGTM overall, I guess we can finish the review in later PRs.
